target/bin/addmailuser

```diff
@@ -73,3 +73,17 @@ then
     sleep 1
   done
 fi
+
+# Per user storage encryption
+if ! grep '^#.*auth-passwdfile-mailcrypt' /etc/dovecot/conf.d/10-auth.conf &>/dev/null
+then
+  while [[ $(stat -c '%U' "/var/mail/${DOMAIN}/${USER}") != docker ]] # handle slow ownership setting problem
```

Comment:
From the commit message:

```
- EFS/slow disk fix for add user

[ec2-user@ip-172-31-3-247 ~]$ ./setup.sh email add -g test3@pierce.us
Enter Password:
Waiting for dovecot to create /var/mail/pierce.us/test3...
total 8.0K
drwxr-sr-x 2 root   docker 6.0K Jul 12 17:41 .
drwxr-sr-x 6 docker docker 6.0K Jul 12 17:41 ..
doveadm(test3@pierce.us): Error: Mailbox INBOX: mkdir(/var/mail/pierce.us/test3/cur) failed: Permission denied (euid=5000(docker) egid=5000(docker) missing +w perm: /var/mail/pierce.us/test3, dir owned by 0:5000 mode=0755)
doveadm(test3@pierce.us): Error: mail_crypt_user_get_public_key(test3@pierce.us) failed: mailbox_open(INBOX) failed: Mailbox INBOX: mkdir(/var/mail/pierce.us/test3/cur) failed: Permission denied (euid=5000(docker) egid=5000(docker) missing +w perm: /var/mail/pierce.us/test3, dir owned by 0:5000 mode=0755)
doveadm(test3@pierce.us): Warning: mailbox cryptokey generate: Nothing was matched. Use -U or specify mask?
doveadm(test3@pierce.us): Panic: file mail-user.c: line 229 (mail_user_deinit): assertion failed: ((*user)->refcount == 1)
doveadm(test3@pierce.us): Error: Raw backtrace: /usr/lib/dovecot/libdovecot.so.0(backtrace_append+0x3d) [0x7f08385083fd] -> /usr/lib/dovecot/libdovecot.so.0(backtrace_get+0x1e) [0x7f083850851e] -> /usr/lib/dovecot/libdovecot.so.0(+0xfcc8b) [0x7f0838514c8b] -> /usr/lib/dovecot/libdovecot.so.0(+0xfccc1) [0x7f0838514cc1] -> /usr/lib/dovecot/libdovecot.so.0(+0x533e7) [0x7f083846b3e7] -> /usr/lib/dovecot/libdovecot-storage.so.0(+0x4065f) [0x7f083861c65f] -> doveadm(+0x32b05) [0x5572a7df7b05] -> doveadm(+0x336da) [0x5572a7df86da] -> doveadm(doveadm_cmd_ver2_to_mail_cmd_wrapper+0x232) [0x5572a7df94a2] -> doveadm(doveadm_cmd_run_ver2+0x4ff) [0x5572a7e09e1f] -> doveadm(doveadm_cmd_try_run_ver2+0x37) [0x5572a7e09e77] -> doveadm(main+0x1ca) [0x5572a7de88ba] -> /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xeb) [0x7f08380a109b] -> doveadm(_start+0x2a) [0x5572a7de8e4a]
/usr/local/bin/addmailuser: line 106:  4254 Aborted                 doveadm -o plugin/mail_crypt_private_password="$(echo -n "${PASSWD}" | sha256sum | awk '{print $1}')" mailbox cryptokey generate -u "${FULL_EMAIL}" -U

YET:

[ec2-user@ip-172-31-3-247 ~]$ docker exec -it deliver bash -c "ls -laht /var/mail/pierce.us/test3"
total 8.0K
drwxr-sr-x 2 docker docker 6.0K Jul 12 17:41 .
drwxr-sr-x 6 docker docker 6.0K Jul 12 17:41 ..

---

Further investigation shows there is a needed sleep for slow disks to safely get the CHOWNed directory:

ec2-user@ip-172-31-3-247 ~]$ ./setup.sh email add -g test3@pierce.us
Enter Password:
Waiting for dovecot to create /var/mail/pierce.us/test3...
Waiting for dovecot to create /var/mail/pierce.us/test3...
Waiting for dovecot to create /var/mail/pierce.us/test3...
Waiting for dovecot to create /var/mail/pierce.us/test3...
Waiting for dovecot to create /var/mail/pierce.us/test3...
Waiting for dovecot to create /var/mail/pierce.us/test3...
Waiting for dovecot to create /var/mail/pierce.us/test3...
Waiting for dovecot to create /var/mail/pierce.us/test3...
Waiting for dovecot to create /var/mail/pierce.us/test3...
Waiting for dovecot to create /var/mail/pierce.us/test3...
Waiting for dovecot to create /var/mail/pierce.us/test3...
Waiting for dovecot to create /var/mail/pierce.us/test3...
Waiting for dovecot to create /var/mail/pierce.us/test3...
total 8.0K
drwxr-sr-x 2 root   docker 6.0K Jul 12 17:44 .
drwxr-sr-x 4 docker docker 6.0K Jul 12 17:44 ..
----- CODE SLEEP FOR 3 SECONDS ------
total 8.0K
drwxr-sr-x 2 docker docker 6.0K Jul 12 17:44 .
drwxr-sr-x 4 docker docker 6.0K Jul 12 17:44 ..
   Folder    Public ID
✓  <userkey> b18c3a7c2f25a88699900a94a53ff5732b865089cd5ad0166fbc9cb03c86f3f5
Successfully created test3@pierce.us

---

I added a while loop which now shows that it's properly waiting:

[ec2-user@ip-172-31-3-247 ~]$ ./setup.sh email del -y test3@pierce.us; ./setup.sh email add -g test3@pierce.us test123
test3@pierce.us and potential aliases deleted.
Mailbox directory '/var/mail/pierce.us/test3' did not exist.
Error :: See the messages above.
Aborting.

--- UNCHECKED ERROR
  - script    = setup.sh
  - function  = _docker_container / ${CRI} exec "${USE_TTY}" "${CONTAINER_NAME}" "${@:+$@}"
  - line      = 253
  - exit code = 1

Make sure you use a version of this script that matches
the version / tag of docker-mailserver. Please read the
'Get the tools' section in the README on GitHub careful-
ly and use ./setup.sh help and read the VERSION section.
Waiting for dovecot to create /var/mail/pierce.us/test3...
Waiting for proper owner to be set on /var/mail/pierce.us/test3...
   Folder    Public ID
✓  <userkey> 3d6a5d196b85627a791532728de8f60ea340e5285c9dc01904c55fa2bdbdfb72
Successfully created test3@pierce.us
[ec2-user@ip-172-31-3-247 ~]$ ./setup.sh email del -y test3@pierce.us; ./setup.sh email add -g test3@pierce.us test123
test3@pierce.us and potential aliases deleted.
Mailbox deleted.
Waiting for dovecot to create /var/mail/pierce.us/test3...
   Folder    Public ID
✓  <userkey> 1d12c659f624a510a3f6bf2b60edc8f1ea61db0fa87742a10fdf0125bc8a39ba
Successfully created test3@pierce.us
```

Reply:
Is there a TL;DR version of that though? It doesn't look like it's got any markdown formatting to separate out the actual text communicating the error vs shell output and commands?

Reply:
The issue was caused by my changes, so I'm not sure I have an issue to reference. Do you recommend I comment this PR?

Reply:
> Do you recommend I comment this PR?

Yes, create a comment here to link to so anyone that tries to alter code understands what the "slow ownership setting problem" is.

It's unlikely someone will bother looking into this PR, through all comments and commit messages to try identify the reasoning? I'm sure it would be more appreciated if you could summarize the problem for future maintainers to understand.

---

I will approve regardless of the issue being documented, but it seems like knowledge of the issue would get lost?

If I understood correctly, performing `chown` can be slow depending on disk latency which can trigger a race condition when the operation is assumed to have already been applied but has not completed. Your solution is to use a while loop that checks every 1 second to confirm successful ownership change?

If that is enough information to understand the problem that the workaround is addressing that can just be an inline comment:

```sh
# avoid race condition with slow disk latency by waiting for chown to complete successfully.
```

If your related commit message provides more insights worth communicating to others, expand on that.